### PR TITLE
JIT: fix handling of PC in dispatcher/block cache.

### PR DESCRIPTION
**Pull request by magumagu**
Specifically, don't make any assumptions about what effective addresses
are used for code, and correctly handle changes to MSR.DR/MSR.IR.

(Split off from dynamic-bat.)

I've only done very basic testing so far, and no performance testing.  ARM needs some additional changes to support this.  Stills need some sort of handling for tlbie.

### CODE REVIEW

File: Source/Core/Core/PowerPC/MMU.cpp [hunk 1]
```diff
@@ -721,6 +729,44 @@ void ClearCacheLine(const u32 address)
...
+		if ((segment == 0x8 || segment == 0x0) && (address & 0x0FFFFFFF) < Memory::REALRAM_SIZE)
+			address = address & 0x3FFFFFFF;
+		else if (segment == 0x9 && (address & 0x0FFFFFFF) < Memory::EXRAM_SIZE)
+			address = address & 0x3FFFFFFF;
+		else if (Memory::bFakeVMEM && (segment == 0x7 || segment == 0x4))
+			address = 0x7E000000 | (address & Memory::FAKEVMEM_MASK);
+		else
+			return TranslateResult{ false, false, 0 };
+	}
+
```
Comment (phire) [inline Source/Core/Core/PowerPC/MMU.cpp]: Oh no, not a stray tab.

How will lint ever survive?

**Comment by tueidj** [comment]
tlbie doesn't translate addresses.

**Comment by magumagu** [comment]
The tlbie issue is that it invalidates the effective->physical address mapping.  This leads to two problems: one, it makes the effective address->block number table invalid, and two, we have to redo block linking computations, which are based on effective addresses.  The first is easy enough to deal with; we can just clear the table.  I still haven't decided what to do about the second.

File: Source/Core/Core/PowerPC/JitCommon/JitCache.h [hunk 2]
```diff
@@ -66,7 +54,7 @@ class ValidBlockBitSet final
 {
 	enum
 	{
-		VALID_BLOCK_MASK_SIZE = 0x20000000 / 32,
+		VALID_BLOCK_MASK_SIZE = 0x100000000 / 32,
```
Comment (FioraAeterna) [inline Source/Core/Core/PowerPC/JitCommon/JitCache.h]: I just hope we're not making the memory usage of our JIT cache too much higher here, what's the overall effect of this patch on that? It seemed pretty wasteful already (I don't fully grok it so I can't judge)

Comment (magumagu) [inline Source/Core/Core/PowerPC/JitCommon/JitCache.h]: This line in particular means the m_valid_block array takes up 16MB; that might be bigger than it needs to be, but making it smaller has complications.

If my computations are correct, this patch saves about 100MB of memory overall.

**Comment by tueidj** [comment]
That sounds bad, if you're relying on tlbie to always be used when an individual effective address mapping is removed it won't work. Case in point: as a workaround for no tlbia instruction games will loop 64 times on tlbie starting with address 0 and incrementing by the page size. So the addresses used are completely bogus, but they still purge the entire TLB regardless (which is all that matters for real hardware).

**Comment by magumagu** [comment]
I'll keep that in mind.  Still have to react to tlbie somehow.

**Comment by tueidj** [comment]
I think you'll have to react to addresses being evicted from the TLB instead.

**Comment by magumagu** [comment]
That's a much better idea than what I was planning; thanks for the suggestion.  (Still not sure how to deal with the block linking issues.)

**Comment by magumagu** [comment]
@dolphin-emu-bot rebuild

**Comment by nbohr1more** [comment]
@dolphin-emu-bot rebuild

**Comment by delroth** [comment]
No activity on the PR and @magumagu is not actively contributing anymore. Closing.

**Pull request closed without merging**